rust/kernel/static_call.rs

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: GPL-2.0
+
+// Copyright (C) 2024 Google LLC.
+
+//! Logic for static calls.
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! to_underscore {
+    ($arg:expr) => {
+        _
+    };
+}
+
+/// Imports the given static call.
+///
+/// To [`static_call!`] a function, you must import it into the local scope using this macro.
+///
+/// [`static_call!`]: macro@static_call
+#[macro_export]
+macro_rules! use_static_call {
+    ($name:ident) => {
+        $crate::macros::paste! { use $crate::bindings:: [<__SCT__ $name >]; }
+        $crate::macros::paste! { use $crate::bindings:: [<__SCK__ $name >]; }
+    };
+}
+
+#[doc(hidden)]
+pub struct AddressableStaticCallKey {
```

Comment:
`#[repr(transparent)]`? This is getting put in a special section, so I suspect having a different layout would break things.

Reply:
I'll add the repr, but technically speaking I don't think it's necessary. It's just used to make sure that the `__SCK__` global does not get stripped from the symbol table.
```c
/*
 * __ADDRESSABLE() is used to ensure the key symbol doesn't get stripped from
 * the symbol table so that objtool can reference it when it generates the
 * .static_call_sites section.
 */
#define __STATIC_CALL_ADDRESSABLE(name) \
	__ADDRESSABLE(STATIC_CALL_KEY(name))
```
The actual logic for patching the call is done by having objtool inspect the binary and use the special names `__SCT__` and `__SCK__` to find the relevant symbols. This static is not used to find them.